### PR TITLE
(Buf)Transients(Slice) Family

### PR DESCRIPTION
**Pull request by tedmoore**

### CODE REVIEW

File: doc/TransientSlice.rst [hunk 1]
```diff
@@ -5,51 +5,54 @@
 :see-also: BufTransientSlice, AmpSlice, NoveltySlice, OnsetSlice
 :description: A real-time transient-based slice extractor
 :discussion: 
-   This uses same algorithm as BufTransients using clicks/transients/derivation/anomaly in the signal to estimate the slicing points.
 
-   The process will return an audio steam with sample-long impulses at estimated starting points of the different slices.
+  TransientSlice identifies slice points in a real time signal by implementing a "de-clicking" algorithm based on the assumption that a transient is a sample or series of samples that are anomalous when compared to surrounding material. It creates an autoregressive model of the samples' time series, so that when a given sample doesn't fit the model (it's "error" or anomalous-ness goes above ``threshFwd``) it is determined to be a transient and a slice point is identified.
```
Comment (jamesb93) [inline doc/TransientSlice.rst:9]: anomalous when compared to surrounding samples

Comment (jamesb93) [reply doc/TransientSlice.rst:9]: Not sure samples' should be possessive about the time series. I would say It creates an autoregressive model of the samples as a time series.

File: doc/TransientSlice.rst [hunk 2]
```diff
@@ -5,51 +5,54 @@
...
 
-   The process will return an audio steam with sample-long impulses at estimated starting points of the different slices.
+  TransientSlice identifies slice points in a real time signal by implementing a "de-clicking" algorithm based on the assumption that a transient is a sample or series of samples that are anomalous when compared to surrounding material. It creates an autoregressive model of the samples' time series, so that when a given sample doesn't fit the model (it's "error" or anomalous-ness goes above ``threshFwd``) it is determined to be a transient and a slice point is identified. 
 
-:process: The audio rate version of the object.
-:output: An audio stream with impulses at detected transients. The latency between the input and the output is (blockSize + padSize - order) samples.
+  The series of samples determined to be a transient will continue until the error goes below ``threshBack``, indicating that the samples are again more in-line with the autoregressive model.
+
+  The process will return an audio steam with sample-long impulses at estimated starting points of the different slices.
```
Comment (jamesb93) [inline doc/TransientSlice.rst:13]: sample-long > single sample

File: doc/TransientSlice.rst [hunk 3]
```diff
@@ -5,51 +5,54 @@
...
+:output: An audio stream with impulses at detected transients. The latency between the input and the output is ``(blockSize + padSize - order)`` samples.
 
 :control in:
 
    The audio to be processed.
 
 :control order:
 
-   The order in samples of the impulse response filter used to model the estimated continuous signal. It is how many previous samples are used by the algorithm to predict the next one as reference for the model. The higher the order, the more accurate is its spectral definition, not unlike fft, improving low frequency resolution, but it differs in that it is not connected to its temporal resolution.
+  How many previous samples are used by the algorithm to create the autoregressive model of the signal within the ``blockSize`` window of analysis.. The higher the ``order``, the more accurate is its spectral definition, improving low frequency resolution (similar to an FFT but it differs in that it is not connected to temporal resolution). ``order`` must be less than ``blockSize``.
```
Comment (jamesb93) [inline doc/TransientSlice.rst:26]: How > The number of

**Review by jamesb93 — CHANGES_REQUESTED** [review]
Some particular comments from me but I think the examples are good. Looking forward to porting them to the Max realm.

File: doc/BufTransientSlice.rst [hunk 4]
```diff
@@ -5,69 +5,73 @@
 :see-also: TransientSlice, BufOnsetSlice, BufNoveltySlice
 :description: Transient-based slice extractor on buffers
 :discussion: 
-   This relies on the same algorithm as BufTransients using clicks/transients/derivation/anomalies in the signal to estimate the slicing points.
 
-   The process will return a buffer which contains indices (in sample) of estimated starting points of the different slices.
+  BufTransientSlice identifies slice points in a buffer by implementing a "de-clicking" algorithm based on the assumption that a transient is a sample or series of samples that are anomalous when compared to surrounding material. It creates an autoregressive model of the samples' time series, so that when a given sample doesn't fit the model (it's "error" or anomalous-ness goes above ``threshFwd``) it is determined to be a transient and a slice point is identified.
```
Comment (weefuzzy) [inline doc/BufTransientSlice.rst:9]: it's "error" -> its "error"

File: doc/BufTransientSlice.rst [hunk 5]
```diff
@@ -5,69 +5,73 @@
 :see-also: TransientSlice, BufOnsetSlice, BufNoveltySlice
 :description: Transient-based slice extractor on buffers
 :discussion: 
-   This relies on the same algorithm as BufTransients using clicks/transients/derivation/anomalies in the signal to estimate the slicing points.
 
-   The process will return a buffer which contains indices (in sample) of estimated starting points of the different slices.
+  BufTransientSlice identifies slice points in a buffer by implementing a "de-clicking" algorithm based on the assumption that a transient is a sample or series of samples that are anomalous when compared to surrounding material. It creates an autoregressive model of the samples' time series, so that when a given sample doesn't fit the model (it's "error" or anomalous-ness goes above ``threshFwd``) it is determined to be a transient and a slice point is identified. 
+  
+  The series of samples determined to be a transient will continue until the error goes below ``threshBack``, indicating that the samples are again more in-line with the autoregressive model.
```
Comment (weefuzzy) [inline doc/BufTransientSlice.rst:11]: To be clear, there are *two* error signals here, forward and backward, so this isn't just a schmitt trigger on a single curve.

File: doc/BufTransientSlice.rst [hunk 6]
```diff
@@ -5,69 +5,73 @@
...
 
 :control indices:
 
-   The index of the buffer where the indices (in sample) of the estimated starting points of slices will be written. The first and last points are always the boundary points of the analysis.
+   The buffer where the indices (in samples) of the estimated starting points of slices will be written.
 
 :control order:
 
-   The order in samples of the impulse response filter used to model the estimated continuous signal. It is how many previous samples are used by the algorithm to predict the next one as reference for the model. The higher the order, the more accurate is its spectral definition, not unlike fft, improving low frequency resolution, but it differs in that it is not conected to its temporal resolution.
+  How many previous samples are used by the algorithm to create the autoregressive model of the signal within the ``blockSize`` window of analysis.. The higher the ``order``, the more accurate is its spectral definition, improving low frequency resolution (similar to an FFT but it differs in that it is not connected to temporal resolution). ``order`` must be less than ``blockSize``.
```
Comment (weefuzzy) [inline doc/BufTransientSlice.rst:46]: Not sure that the spectral resolution stuff is illuminating here, but yes, higher order models will cope better with more distal temporal dependencies in data, which translates to handling low frequencies better

File: doc/BufTransientSlice.rst [hunk 7]
```diff
@@ -5,69 +5,73 @@
...
 
 :control skew:
 
-   The nervousness of the bespoke detection function with values from -10 to 10. It allows to decide how peaks are amplified or smoothed before the thresholding. High values increase the sensitivity to small variations.
+ The nervousness of the bespoke detection function. It ranges from -10 to 10 (it has no units) representing the strength and direction of some nonlinearity applied to the detection signal which controls how peaks are amplified or smoothed before the thresholding. High values increase the sensitivity to small variations.
 
 :control threshFwd:
 
-   The threshold of the onset of the smoothed error function. It allows tight start of the identification of the anomaly as it proceeds forward.
+ The threshold applied to the smoothed error function for determining an onset. The units are roughly in standard deviations, thus can be considered how "deviant", or anomalous, the signal must be to be detected as a transient. It allows tight start of the identification of the anomaly as it proceeds forward.
```
Comment (weefuzzy) [inline doc/BufTransientSlice.rst:62]: smoothed forward error function

Comment (weefuzzy) [reply doc/BufTransientSlice.rst:62]: or even forward prediction error

File: doc/BufTransientSlice.rst [hunk 8]
```diff
@@ -5,69 +5,73 @@
...
 
 :control threshBack:
 
-   The threshold of the offset of the smoothed error function. As it proceeds backwards in time, it allows tight ending of the identification of the anomaly.
+ The threshold applied to the smoothed error function for determining an offset. The units are roughly in standard deviations, thus can be considered how "deviant", or anomalous, the signal must be to be considered transient. When the smoothed error function goes below ``threshBack`` an offset is identified. As it proceeds backwards in time, it allows tight ending of the identification of the anomaly.
```
Comment (weefuzzy) [inline doc/BufTransientSlice.rst:66]: ditto backward

File: doc/BufTransientSlice.rst [hunk 9]
```diff
@@ -76,4 +80,3 @@
 :control action:
```
Comment (weefuzzy) [inline doc/BufTransientSlice.rst:80]: do we need documentation for `action`s. I still can't remember 😄

**Review by weefuzzy — CHANGES_REQUESTED** [review]

File: doc/BufTransientSlice.rst (hunk 5 above)
Comment (tedmoore) [reply doc/BufTransientSlice.rst:11]: @weefuzzy Yeah, I gathered that that is the case. Is it ok to allow our reader to think of it as Schmitt Trigger that uses these critera as onset and offset detectors? (This explanation can yet change, just trying to not overexplain things)

File: doc/BufTransientSlice.rst [hunk 10]
```diff
@@ -5,75 +5,75 @@
 :see-also: TransientSlice, BufOnsetSlice, BufNoveltySlice
 :description: Transient-based slice extractor on buffers
 :discussion: 
-   This relies on the same algorithm as BufTransients using clicks/transients/derivation/anomalies in the signal to estimate the slicing points.
 
-   The process will return a buffer which contains indices (in sample) of estimated starting points of the different slices.
+  BufTransientSlice identifies slice points in a buffer by implementing a "de-clicking" algorithm based on the assumption that a transient is a sample or series of samples that are anomalous when compared to surrounding samples. It creates an autoregressive model of the time series of samples, so that when a given sample doesn't fit the model (its "error" or anomalous-ness goes above ``threshFwd``) it is determined to be a transient and a slice point is identified.
```
Comment (tremblap) [inline doc/BufTransientSlice.rst:9]: maybe taming down 'autoregressive' by saying just before - it creates a model to predict something similar to its surrounding via autoregression (in proper english - but defering the algo after the explanation of guestimating continuation)

File: doc/BufTransientSlice.rst [hunk 11]
```diff
@@ -5,75 +5,75 @@
...
 
 :control padSize:
 
-   The size of the handles on each sides of the block simply used for analysis purpose and avoid boundary issues.
+  The size (in samples) of analysis on each side of ``blockSize`` used to provide some historical context for analysis so that each ``blockSize`` isn't modelled completely independently of its predecessor.
 
 :control skew:
 
-   The nervousness of the bespoke detection function with values from -10 to 10. It allows to decide how peaks are amplified or smoothed before the thresholding. High values increase the sensitivity to small variations.
+ The nervousness of the bespoke detection function. It ranges from -10 to 10 (it has no units) representing the strength and direction of some nonlinearity applied to the detection signal which controls how peaks are amplified or smoothed before the thresholding. High values increase the sensitivity to small variations.
```
Comment (tremblap) [inline doc/BufTransientSlice.rst:58]: high is ambiguous in high values. positive values?

File: doc/BufTransients.rst [hunk 12]
```diff
@@ -3,81 +3,78 @@
...
 :see-also: Transients, BufHPSS, BufSines
-:description: A transient extractor on buffers
+:description: Separate Transients from a Signal in a Buffer
 :discussion: 
-   It implements declicking algorithm from chapter 5 of 'Digital Audio Restoration' by Godsill, Simon J., Rayner, Peter J.W. with some bespoke improvements on the detection function tracking.
 
-   The algorithm will take a buffer in, and will divide it in two outputs:
-     * the transients, estimated from the signal and extracted from it;
-     * the remainder of the material, as estimated by the algorithm.
+  This implements a "de-clicking" algorithm based on the assumption that a transient is a sample or series of samples that are anomalous when compared to surrounding samples. It creates an autoregressive model of the time series of samples, so that when a given sample doesn't fit the model (its "error" or anomalous-ness goes above ``threshFwd``) it is determined to be a transient. The series of samples determined to be a transient will continue until the error goes below ``threshBack``, indicating that the samples are again more in-line with the autoregressive model.
```
Comment (tremblap) [inline doc/BufTransients.rst:9]: same remark as above in the RT version

File: doc/TransientSlice.rst [hunk 13]
```diff
@@ -5,51 +5,54 @@
 :see-also: BufTransientSlice, AmpSlice, NoveltySlice, OnsetSlice
 :description: A real-time transient-based slice extractor
 :discussion: 
-   This uses same algorithm as BufTransients using clicks/transients/derivation/anomaly in the signal to estimate the slicing points.
 
-   The process will return an audio steam with sample-long impulses at estimated starting points of the different slices.
+  TransientSlice identifies slice points in a real time signal by implementing a "de-clicking" algorithm based on the assumption that a transient is a sample or series of samples that are anomalous when compared to surrounding samples. It creates an autoregressive model of the time series of samples, so that when a given sample doesn't fit the model (its "error" or anomalous-ness goes above ``threshFwd``) it is determined to be a transient and a slice point is identified.
```
Comment (tremblap) [inline doc/TransientSlice.rst:9]: and here again

File: doc/Transients.rst [hunk 14]
```diff
@@ -3,53 +3,54 @@
...
 
-   The algorithm will take an audio in, and will divide it in two outputs:
-   	* the transients, estimated from the signal and extracted from it;
-   	* the remainder of the material, as estimated by the algorithm.
-
-   	The whole process is based on the assumption that a transient is an element that is deviating from the surrounding material, as sort of click or anomaly. The algorithm then estimates what should have happened if the signal had followed its normal path, and resynthesises this estimate, removing the anomaly which is considered as the transient.
+   This implements a "de-clicking" algorithm based on the assumption that a transient is a sample or series of samples that are anomalous when compared to surrounding samples. It creates an autoregressive model of the time series of samples, so that when a given sample doesn't fit the model (its "error" or anomalous-ness goes above ``threshFwd``) it is determined to be a transient. The series of samples determined to be a transient will continue until the error goes below ``threshBack``, indicating that the samples are again more in-line with the autoregressive model.
```
Comment (tremblap) [inline doc/Transients.rst:9]: and here. I wonder if there isn't a way to not copy that text 4 times.

File: example-code/sc/Transients.scd [hunk 15]
```diff
@@ -1,14 +1,56 @@
...
-// basic parameters
-{FluidTransients.ar(PlayBuf.ar(1, b, loop:1))}.play
+~src = Buffer.read(s,FluidFilesPath("Tremblay-AaS-SynthTwoVoices-M.wav"));
 
-// tweaked parameters
-{FluidTransients.ar(PlayBuf.ar(1, b, loop:1), 80, threshFwd:MouseX.kr(0,5), threshBack:MouseY.kr(0,2))}.play
+// hear the basic parameters: transients in the L channel, residual in the R channel
+(
+{
+	FluidTransients.ar(PlayBuf.ar(1, b, loop:1));
```
Comment (tremblap) [inline example-code/sc/Transients.scd:9]: bug: reading b and not ~src

File: example-code/sc/Transients.scd [hunk 16]
```diff
@@ -1,14 +1,56 @@
...
+	FluidTransients.ar(sig,order,skew:skew)[0].dup
+}.play;
+)
+
+::
+strong::adding some FX::
+code::
+(
+{
+	var sig = PlayBuf.ar(1, b, loop:1);
```
Comment (tremblap) [inline example-code/sc/Transients.scd:28]: same bug not playing ~src

File: example-code/sc/BufTransients.scd [hunk 17]
```diff
@@ -1,68 +1,53 @@
...
-    (Main.elapsedTime - t).postln;
-}.play
+~transients = Buffer(s);
+~residual = Buffer(s);
+~compare = Buffer(s);
+FluidBufTransients.processBlocking(s,~src,transients:~transients,residual:~residual,order:18,skew:10,threshFwd:1.1);
+FluidBufCompose.processBlocking(s,~src,destination:~compare,destStartChan:0);
+FluidBufCompose.processBlocking(s,~transients,destination:~compare,destStartChan:1);
+FluidBufCompose.processBlocking(s,~residual,destination:~compare,destStartChan:2);
+FluidWaveform(~compare,bounds:Rect(0,0,1600,1200));
```
Comment (tremblap) [inline example-code/sc/BufTransients.scd:51]: maybe zooming in to see the difference?

**Review by tremblap — CHANGES_REQUESTED** [review]
a great explanation, a few proposals, and a few bugs.

File: example-code/sc/BufTransients.scd (hunk 17 above)
Comment (tedmoore) [reply example-code/sc/BufTransients.scd:51]: FluidWaveform doesn't have zoom yet.

Comment (tremblap) [reply example-code/sc/BufTransients.scd:51]: I know but one could use 1 second of source sound instead and we can see the wave...

**Pull request merged**